dpctl/_sycl_device.pyx

```diff
@@ -157,6 +188,87 @@ cdef class SyclDevice(_SyclDevice):
+        DPCTLDeviceVector_Delete(DVRef)
+        return devices
+
+    cdef list create_sub_devices_by_counts(self, list counts):
+        """ Returns a vector of sub devices
+            partitioned from this SYCL device based on the counts parameter. For each
+            non-zero value M in the counts vector, a sub device with M compute units
+            is created.
+        """
+        cdef size_t ncounts = len(counts)
+        cdef size_t *counts_buff = <size_t *> malloc(ncounts * sizeof(size_t))
+        cdef DPCTLDeviceVectorRef DVRef = NULL
+        cdef int i
+        for i in range(ncounts):
+            counts_buff[i] = counts[i]
+        DVRef = DPCTLDevice_CreateSubDevicesByCounts(self._device_ref, counts_buff, ncounts)
+        if DVRef is NULL:
+            self._raise_sub_devices_creation_error("DPCTLSubDeviceCreationError", -1)
+        cdef list devices = _get_devices(DVRef)
+        free(counts_buff)
+        DPCTLDeviceVector_Delete(DVRef)
+        return devices
+
+    cdef list create_sub_devices_by_affinity(self, _partition_affinity_domain_type domain):
+        """ Returns a vector of sub devices
+            partitioned from this SYCL device by affinity domain based on the domain
+            parameter.
+        """
+        cdef DPCTLDeviceVectorRef DVRef = NULL
+        DVRef = DPCTLDevice_CreateSubDevicesByAffinity(self._device_ref, domain)
+        if DVRef is NULL:
+            self._raise_sub_devices_creation_error("DPCTLSubDeviceCreationError", -1)
+        cdef list devices = _get_devices(DVRef)
+        DPCTLDeviceVector_Delete(DVRef)
+        return devices
+
+    def create_sub_devices(self, partition):
+        if isinstance(partition, int) and partition > 0:
+            return self.create_sub_devices_equally(partition)
+        elif isinstance(partition, list) and all([i > 0 for i in partition]):
+            return self.create_sub_devices_by_counts(partition)
+        elif isinstance(partition, str):
+            if partition == "not_applicable":
+                domain_type = _partition_affinity_domain_type._not_applicable
+            elif partition == "numa":
+                domain_type = _partition_affinity_domain_type._numa
+            elif partition == "L4_cache":
+                domain_type = _partition_affinity_domain_type._L4_cache
+            elif partition == "L3_cache":
+                domain_type = _partition_affinity_domain_type._L3_cache
+            elif partition == "L2_cache":
+                domain_type = _partition_affinity_domain_type._L2_cache
+            elif partition == "L1_cache":
+                domain_type = _partition_affinity_domain_type._L1_cache
+            elif partition == "next_partitionable":
+                domain_type = _partition_affinity_domain_type._next_partitionable
+            else:
+                raise Exception('Unsupported type of domain')
+            return self.create_sub_devices_by_affinity(domain_type)
+        else:
```

Comment:
I reworked this dispatch in #343 to require keyword `partition`, so that `device.create_sub_devices(4)` must become `device.create_sub_devices(partition=4)` to work.

Changes allow allow for list/tuples/numpy arrays for specify partition by counts.

```

In [1]: import dpctl, numpy as np

In [2]: [di.max_compute_units for di in dpctl.SyclDevice("cpu").create_sub_devices(partition=np.array([2, 6])) ]
Out[2]: [2, 6]

In [1]: import dpctl, numpy as np

In [2]: [di.max_compute_units for di in dpctl.SyclDevice("cpu").create_sub_devices(partition=np.array([2, 6])) ]
Out[2]: [2, 6]

In [4]: [di.max_compute_units for di in dpctl.SyclDevice("cpu").create_sub_devices(partition = [2, 6, 4]) ]
Out[4]: [2, 6, 4]

In [5]: [di.max_compute_units for di in dpctl.SyclDevice("cpu").create_sub_devices(partition = (2, 6, 4)) ]
Out[5]: [2, 6, 4]
```

Also use of NumPy integral types is supported:

```
In [6]: dpctl.SyclDevice("cpu").create_sub_devices(partition=np.int32(4))
Out[6]:
[<dpctl.SyclDevice [backend_type.opencl, device_type.cpu,  Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz] at 0x7f1f698727b0>,
 <dpctl.SyclDevice [backend_type.opencl, device_type.cpu,  Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz] at 0x7f1f69872c30>,
 <dpctl.SyclDevice [backend_type.opencl, device_type.cpu,  Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz] at 0x7f1f69872d30>]
```